src/librustc_data_structures/sharded.rs

```diff
@@ -60,6 +60,11 @@ impl<T> Sharded<T> {
         }
     }
 
+    #[inline]
```

Comment:
This needs a comment describing how the indices are distributed across the shards -- or, better yet, maybe it could return a `(&Lock<T>, usize)`, where the usize is the index within the shard. In particular, I was surprised at first to see that we selected the shard via modulo and the index *within* the shared via `/`, though it makes sense I guess (put adjacent nodes far apart seems like a reasonable heuristic).

I guess returning a tuple would make the `LockGuard` code above a bit more complex, but it would allow you to encapsulate both the `%` and the `/` in one function, which seems strictly better.